beacon-chain/db/kv/state.go

```diff
@@ -318,3 +320,44 @@ func createStateIndicesFromStateSlot(ctx context.Context, slot uint64) map[strin
 	}
 	return indicesByBucket
 }
+
+// CleanUpDirtyStates removes states in DB that falls to under archived point interval rules.
+// Only following states would be kept:
+// 1.) state_slot % archived_interval == 0. (e.g. archived_interval=2048, states with slot 2048, 4096... etc)
+// 2.) archived_interval - slots_per_epoch/2 < state_slot % archived_interval
+//   (e.g. archived_interval=2048, slots_per_epoch=32, states with slot 2047, 2046, 2032... etc).
+//   This is to tolerate skip slots. Not every state lays on the boundary.
+// 3.) state with current finalized root
+func (s *Store) CleanUpDirtyStates(ctx context.Context, slotsPerArchivedPoint uint64) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB. CleanUpDirtyStates")
+	defer span.End()
+	f, err := s.FinalizedCheckpoint(ctx)
+	if err != nil {
+		return err
+	}
+	deletedRoots := make([][32]byte, 0)
+
+	err = s.db.View(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(stateSlotIndicesBucket)
+		if err := bkt.ForEach(func(k, v []byte) error {
+			finalized := bytesutil.ToBytes32(f.Root) == bytesutil.ToBytes32(v)
+			slot := bytesutil.BytesToUint64BigEndian(k)
+			mod := slot % slotsPerArchivedPoint
+			// The following conditions cover 1, 2, and 3 above.
+			if mod != 0 && mod <= slotsPerArchivedPoint-params.BeaconConfig().SlotsPerEpoch/2 && !finalized {
```

Comment:
if `slotsPerArchivedPoint < params.BeaconConfig().SlotsPerEpoch/2` , then you have an overflow.

Reply:
Also still not clear on this particular chunk of logic
```
 slotsPerArchivedPoint-params.BeaconConfig().SlotsPerEpoch/2 
```
what if the whole epoch is skipped then ? you could have the last checkpoint at 2015 instead. You would miss a boundary state here and delete it.

Reply:
Right. This check is really best effort. What number would you feel better here if not half of `slotsPerEpoch`?  half of `slotsPerArchivedPoint`?

Reply:
Anyway we can keep track of the previous boundary state and then follow on from there ? Assuming the slot keys are lexicographically sorted, we can keep track of the previous valid boundary state and then follow on from there. Although this might increase complexity of this routine, you would have something like this:
```
previous_boundary_state 
state A
state B
state C 
---next_boundary
```
You would have to keep track of A,B,C till we hit the next boundary and then delete A, B . I dont feel too strongly here, but given that we might prematurely delete a valid state, any user querying historical states would have to start 2 * 2048 slots behind.

Reply:
I was using medalla for example. This would have happened only twice.

With that said, I think i'd feel better on the value `slotsPerArchivedPoint / 3`.  It lessen the chance of it happening. Let me switch to that and test medalla again

Reply:
Can confirm, I like `slotsPerArchivedPoint / 3` more. It only deleted 190 out of 220 states but it's much safer. I'd rather under delete than over delete here